### ISSUE TITLE
LoggerConfiguration.CreateLogger() should disallow multiple calls

### ISSUE THREAD

**Issue by nblumhardt** [serilog/serilog#561]
While `LoggerConfiguration` itself can safely handle `CreateLogger()` being called more than once, many of the sinks are disposable, and disposing one of the loggers may break the others.

To avoid this ugly situation, `CreateLogger()` should throw `InvalidOperationException` with an explanatory message if called more than once.

(Possible cause of #559).

**Comment by daveaglick** [comment]
Fixing this outright with a flag and exception on repeated calls looks to be quite easy. However, if the primary issue here is multiple disposal of sinks, I wonder if there isn't a way to fix the problem without relying on "only call once" semantics. After looking at the code, I can think of two possible ways to address this besides throwing:

- Make the `LoggerSinkConfiguration` store delegates for constructing the sinks instead of actual sink instances. Then when `CreateLogger()` is called, spin up a new instance of each sink for each logger. Of course, this would drastically change the semantics of `LoggerSinkConfiguration` and would require sinks have the ability to store their configuration separate from an instance so that the delegate wouldn't get too messy. There may be other problems with the approach of having multiple sinks that do the same thing (file lock contention, etc.)
- The sinks could "reference count" their attached loggers and only dispose when there are no more loggers left. I like this approach better because it doesn't require changing the public API. However, there's an issue of what should happen if you dispose the first logger out (thus not disposing the sinks) but then never dispose the second/last one? The caller would have to make sure to dispose all loggers to trigger sink disposal (which may not be a big deal if they're interested in disposal in the first place).

Thoughts?

**Comment by nblumhardt** [comment]
Thanks for the thoughts @daveaglick - I guess the flipside to this is, what scenarios would we be seeking specifically to enable by allowing multiple `Configure()` calls?

(I've also seen disastrous performance come from this too - e.g. creating an entirely new logging pipeline per HTTP request and so-on, so I view this as a bit of a usability issue to fix here as well...)

**Comment by daveaglick** [comment]
Yeah, the more I thought about it, the more I started to wonder if I was trying to create a solution without a problem. I do still wonder about the design of having a configuration object that gets "used up" so that instance methods start to throw on multiple invokes. Something about that seems like a smell.

Maybe the right way to address it is through better documentation or a slight redesign of the API so that the creation of a logger and the storage of settings for that operation are separated.

Is this the kind of thing you think warrants some investigation, or does it fall under "get it done and move on" (and there's nothing wrong with the latter - I'm a firm believer that sometimes perfect is the enemy of good enough).

**Comment by nblumhardt** [comment]
I think the simple notion of "resource ownership transfer" between the configuration object and the logger it creates is the best one here. Let's :shipit: and move on :-)

**Comment by jeremymeng** [comment]
@nblumhardt Castle.Core has this test that create different loggers for different contexts https://github.com/castleproject/Core/blob/v3.3.3/src/Castle.Core.Tests/SerilogIntegration/SerilogTests.cs#L163-L183.

What is your recommendation to react to this change?  Thanks!

**Comment by IanYates** [comment]
I should probably make a PR to Castle.  I found the Serilog integration there to be a bit broken and ended up writing my own since it's really only a class and a wrapper.

You can just skip the Castle logger factory stuff and instead just register a simple wrapper class implementing Castle's ILogger, deferring all calls to Serilog's ILogger, in Castle directly.

**Comment by nblumhardt** [comment]
Thanks for the heads-up @jeremymeng. Sounds like @IanYates has more background than I do here.

A few thoughts and obsevations though:

```csharp
            var config = new LoggerConfiguration()
                .MinimumLevel.Debug()
                .WriteTo.TextWriter(output);

            var factory = new SerilogFactory(config);
```

Having `SerilogFactory` accept a `LoggerConfiguration` is a bit of a risky proposition. Things will go awry, for example, if the configuration includes something like `WriteTo.File()` where at most one configuration can use the specified file.

If instead, `SerilogFactory` accepted an `ILogger`, customised loggers for various contexts can be created from it:

```csharp
var componentA = log.ForContext("Component", "A");
componentA.Information("Hello from A!");

// or

var componentB = log.ForContext(typeof(ComponentB));
```

There's an example Autofac integration for Serilog here that could provide some inspiration: https://github.com/nblumhardt/autofac-serilog-integration

Let me know if this makes sense or if I can provide more info.

**Comment by IanYates** [comment]
The suggestion from @nblumhardt is spot on and that's the change to make to Windsor.  The trick is of course that it's a breaking change (although the current approach is broken in subtle ways anyway).

The customised context loggers is exactly how I did it.

I'll make that PR but I'm pretty flat out for a couple of days so unless I can sneak it in after the kids go to bed it probably won't be until Thursday later this week.

**Comment by jonorossi** [comment]
I merged the PR to Castle Core contributed by @ruba1987 quite a long time ago without much knowledge of Serilog, I've made a few bug fixes since the initial PR, but I don't think many people are using it, if it needs to change that's not a problem.

@IanYates no problem with being a breaking change, we are upgrading to Serilog 2.0 for v4 and we can push the new factory out in Castle Core v4. Happy to see a PR for this in the next few weeks, I've created https://github.com/castleproject/Core/issues/142 to make sure we get this fixed for v4.
